packages/sdk/src/oracle/utils.ts

```diff
@@ -0,0 +1,39 @@
+import { State } from "./types/state";
+import { Read } from "./store";
+
+export enum RequestState {
+  Invalid = 0, // Never requested.
+  Requested, // Requested, no other actions taken.
+  Proposed, // Proposed, but not expired or disputed yet.
+  Expired, // Proposed, not disputed, past liveness.
+  Disputed, // Disputed, but no DVM price returned yet.
+  Resolved, // Disputed and DVM price is available.
+  Settled, // Final price has been set in the contract (can get here from Expired or Resolved).
+}
+
+export enum ProposalFlags {
+  WrongChain = 0,
+  InvalidContractState,
+  InsufficientBalance,
+  InsufficientApproval,
+  TransactionInProgress,
+}
+export function previewProposal(state: State) {
```

Comment:
Should we also simulate the txn and add an error to signal that the txn failed (for an unknown reason)?

Reply:
this is really just for showing changes in the proposal form, so a failing transaction will probably be shown elsewhere. i may generalize this in a later pr, but not exactly sure yet how i want to deal with transactions.